src/spaces/compartment.jl

```diff
@@ -0,0 +1,143 @@
+    a.pos = pos
+    add_agent_to_space!(a, model)
+end
+
+"""
+move_agent!(agent::A, model::ABM{A, CompartmentSpace}, dt = 1.0)
+Propagate the agent forwards one step according to its velocity,
+_after_ updating the agent's velocity (see [`CompartmentSpace`](@ref)).
+Also take care of periodic boundary conditions.
+
+For this continuous space version of `move_agent!`, the "evolution algorithm"
+is a trivial Euler scheme with `dt` the step size, i.e. the agent position is updated
+as `agent.pos += agent.vel * dt`.
+
+Notice that if you want the agent to instantly move to a specified position, do
+    `move_agent!(agent, pos, model)`.
+    """
+    function move_agent!(agent::A, model::ABM{A, <: CompartmentSpace}, dt::Real = 1.0) where {A <: AbstractAgent}
+        model.space.update_vel!(agent, model)
+        pos = agent.pos .+ dt .* agent.vel
+        move_agent!(agent, pos, model)
+        return agent.pos
+    end
+    
+    #######################################################################################
+    # %% Neighbors and stuff
+    #######################################################################################
+    
+    grid_space_neighborhood(α, model::ABM{<:AbstractAgent, <:CompartmentSpace}, r) =
+    grid_space_neighborhood(α, model.space.grid, r)
+    
+    function nearby_ids(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r = 1)
+        nn = grid_space_neighborhood(CartesianIndex(pos2cell(pos)), model, r)
+        s = model.space.grid.s
+        Iterators.flatten((s[i...] for i in nn))
+    end
+    
+    function nearby_positions(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r = 1)
+        nn = grid_space_neighborhood(CartesianIndex(pos2cell(pos)), model, r)
+        Iterators.filter(!isequal(pos), nn)
+    end
+    
+    function positions(model::ABM{<:AbstractAgent,<:CompartmentSpace})
+        x = CartesianIndices(model.space.grid.s)
+        return (Tuple(y) for y in x)
+    end
+    
+    function ids_in_position(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace})
+        return model.space.grid.s[pos...]
+    end
+    
+    cell_center(pos::Tuple) = getindex.(modf.(pos), 2) .+ 0.5
+    distance_from_cell_center(pos::Tuple, center) = sqrt(sum(abs2.(pos .- center)))
+    
+    """
+    space_neighbors(position, model::ABM, r=1; kwargs...) → ids
+    
+    Return an iterator of the ids of the agents within "radius" `r` of the given `position`
+    (which must match type with the spatial structure of the `model`).
+    """
+    function space_neighbors(pos, model::ABM{<:AbstractAgent,<:CompartmentSpace}, r=1; exact=false)
+        center = cell_center(pos)
+        focal_cell = ceil.(Int, center)
+        if exact
+            newr = ceil.(Int, r)
+            corner_of_largest_square_in_circle = floor.(Int, center .+ newr)
+            max_distance = corner_of_largest_square_in_circle[1] - focal_cell[1]
+            final_ids = Int[] # TODO make it an iterator
```

Comment:
@Datseris Do you know how to make this an iterator?

Reply:
I'm fairly certain you cannot append to an iterator, so making this one doesn't make sense with this current implementation.